src/common/snippets/src/lowered/pass/extract_loop_invariants.cpp

```diff
@@ -0,0 +1,204 @@
+        loop_manager->sort_loop_ports(inner_loop_begin_pos, inner_loop_end_pos, inner_loop_id);
+    }
+}
+
+std::set<ExpressionPtr> ExtractLoopInvariants::get_potential_extractable_exprs(const std::vector<LoopPort>& loop_in_ports) {
+    std::set<ExpressionPtr> expr_set;
+    for (size_t i = 0; i < loop_in_ports.size(); ++i) {
+        expr_set.insert(loop_in_ports[i].expr_port->get_expr());
+    }
+    return expr_set;
+}
+
+bool ExtractLoopInvariants::extract_from_loop(const size_t& inner_loop_id, LinearIR& linear_ir) {
+    const auto& loop_manager = linear_ir.get_loop_manager();
+    bool status = false;
+    bool extraction_completed = false;  // true means all extractable exprs are extracted from this loop
+    const auto& inner_loop_info = loop_manager->get_loop_info<UnifiedLoopInfo>(inner_loop_id);
+    while (!extraction_completed) {
+        const auto& inner_loop_input_ports = inner_loop_info->get_input_ports();
+        const auto& potential_extractable_exprs = get_potential_extractable_exprs(inner_loop_input_ports);
+        bool has_been_extraction = false;
+        for (const auto& port_expr : potential_extractable_exprs) {
+            if (is_extraction_applicable(port_expr, inner_loop_info)) {
+                status = true;
+                LinearIR::constExprIt inner_loop_begin_pos, inner_loop_end_pos;
+                std::tie(inner_loop_begin_pos, inner_loop_end_pos) =
+                    loop_manager->get_loop_bounds(linear_ir, inner_loop_id);
+                // extract scalar on inputs if there are
+                for (size_t i = 0; i < port_expr->get_input_count(); ++i) {
+                    auto parent = port_expr->get_input_port_connector(i)->get_source().get_expr();
+                    if (ov::is_type<snippets::op::Scalar>(parent->get_node())) {
+                        extract_expr(parent, linear_ir, inner_loop_begin_pos, inner_loop_end_pos);
+                    }
+                }
+                extract_expr(port_expr, linear_ir, inner_loop_begin_pos, inner_loop_end_pos);
+                update_loop_ports(port_expr, loop_manager, inner_loop_id, inner_loop_begin_pos, inner_loop_end_pos);
+                has_been_extraction = true;
+                break;  // extracted and refreshed loop_input_ports. break potential_extractable_exprs loop, and go while() to start again.
+            }
+        }
+        if (inner_loop_input_ports.size() == 0 && inner_loop_info->get_output_ports().size() == 0) {
+            // become a empty(inner_loop_input_ports is ref) loop after extraction, let remove it from loop_manager
+            loop_manager->remove_loop_info(inner_loop_id);
+            extraction_completed = true;
+            break;
+        }
+        // no more extractable expr in this loop after for() of iter loop ports, done for this loop.
+        if (!has_been_extraction)
+            extraction_completed = true;
```

Comment:
We ca eliminate this check entirely if we:
1) add `bool has_been_extraction = true;` before the while
2) change the while condition to `has_been_extraction`
3) modify L137 to simply  has_been_extraction = false

If we do so, we can also get rid of the useless `status` variable and return simply `has_been_extraction`.

I might've missed something, please double-check, but it least that we can simplify this `while` and increase readability

Reply:
@IvanNovoselov , The major algo is:
1. We get `potential_extractable_exprs` (the input port exprs). Go through these exprs, and if one expr could be extracted, extract(move outside) it and update loop port, get new `potential_extractable_exprs` based on updated loop port. Go on in next iteration(while()) with new `potential_extractable_exprs`. In each iteration only at most one expr could be extracted.
2. If none of the `potential_extractable_exprs` is extracted, this means all extractable exprs are already extracted. Extraction is finished for this loop.

Given this process, I applied your suggestion 1) and 2) with `bool continue_to_extract = true` and `expr_extracted` in each iteration, for more readability.
For 3), I think this check` if (!has_been_extraction) extraction_completed = true;` is actually for what you ask. `expr_extracted` is for each iteration, after going through all `potential_extractable_exprs`, if none is extracted, we will set continue_to_extract  to false, which should be done outside of `for(potential_extractable_exprs)` loop. If one is extracted, continue_to_extract is still true, go to next iteration with updated loop port.
As for status, if one expr is extracted in any iteration. It should be ture. `expr_extracted` is for each iteration, not entirely the same value. If 2 exprs extracted, in the 1 and 2 iteration, expr_extracted is ture, in the 3 iteration, it's false, set continue_to_extract  to false and finish the loop extraction. This [comment ](https://github.com/openvinotoolkit/openvino/pull/24044#discussion_r1628830725)also provide some explanation.
Welcome discussion if more suggestions. Thank you Ivan.

Reply:
Oh, so I'm not the first who was confused by this method 🤔  
Ok, if you are sure that it can't be simplified, I'll take your word for it 🙂